controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```
@@ -632,11 +632,30 @@ public void initializeStreamWriters(final EventStreamClientFactory clientFactory
+        // if our update succeeded or not. Posting the event is harmless. If the update had succeeded, then the event will 
+        // be used for processing. If the update had failed, the event will be discarded. We will throw the exception 
+        // that we thrown from running futureSupplier.
+        AtomicReference<Throwable> toThrow = new AtomicReference<>();
         return streamMetadataStore.addRequestToIndex(context.getHostId(), id, event)
-                           .thenCompose(v -> futureSupplier.get())
-                           .thenCompose(t -> RetryHelper.withIndefiniteRetriesAsync(() -> writeEvent(event), e -> { }, executor)
+                           .thenCompose(v -> Futures.exceptionallyExpecting(futureSupplier.get(), 
+                                   e -> {
+                                       toThrow.set(e);
+                                       return Exceptions.unwrap(e) instanceof StoreException.StoreConnectionException ||
+                                               Exceptions.unwrap(e) instanceof StoreException.WriteConflictException;
+                                   }, null))
+                           .thenCompose(t -> RetryHelper.withIndefiniteRetriesAsync(() -> writeEvent(event),
+                                   e -> log.warn("writing event failed with {}", e.getMessage()), executor)
                                                         .thenCompose(v -> streamMetadataStore.removeTaskFromIndex(context.getHostId(), id))
-                                                        .thenApply(v -> t));
+                                                        .thenApply(v -> {
+                                                            if (toThrow.get() != null) {
+                                                                throw new CompletionException(toThrow.get());
+                                                            } else {
+                                                                return t;
```

I don't understand why we are returning `t` here.

---

The method (`addIndexAndSubmitTask`) returns the result of executing futureSupplier which is `t`. 
The original code was also ultimately returning `t` after writing the event. 
With this change, we are "handling" exceptions thrown from executing futureSupplier and as a safety we are writing event in case of StoreException or WriteConflictException thrown from futureSupplier. And in successful case we return its result (which is `t`) and in failure case we throw the exception.

---

Could we refactor this part of the code and use say `exceptionallyComposeExpecting` to avoid having to pass an atomic reference for a throwable outside the completion stage flow?

---

We want to throw the exception even after performing a `thenCompose` on expected failures. This is to indicate to the caller that the requested operation "may" have failed since we received an exception from the store.

---

Can we use `thenCombine` to combine the result of `addRequestToIndex` and the retry loop to remove tasks from index?